File: Python/0114-m-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Non-recursive Preorder Traversal
class Solution_1(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        pointer = dummy = TreeNode(None)
        stack = [root]
        while stack:
            top = stack.pop()
            if not top: continue
            stack.append(top.right)
            stack.append(top.left)
            pointer.right = top
            pointer.left = None
            pointer = top

# Recursive Postorder Traversal
class Solution_2(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        self.pointer = None
        def traverse(root):
            if not root: return
            traverse(root.right)
            traverse(root.left)
            root.right = self.pointer
            root.left = None
            self.pointer = root
        traverse(root)

File: leetcode/C++/0114-m-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Recursion
class Solution {
public:
    void flatten(TreeNode* root) {
        flattenTree(root);
    }
    
    TreeNode* flattenTree(TreeNode* node) {
        if (!node) return nullptr;
        if (!node->left && !node->right) return node;
        TreeNode* leftTail = flattenTree(node->left);
        TreeNode* rightTail = flattenTree(node->right);

        if (leftTail) {
            leftTail->right = node->right;
            node->right = node->left;
            node->left = nullptr;
        }
        
        return rightTail ? rightTail : leftTail;
    }
};


// Stack
class Solution {
public:
    void flatten(TreeNode* root) {
        if (!root) return;
        stack<TreeNode *> stk;
    
        while (1) {
            if (root->right)
                stk.push(root->right);

            if (root->left) {
                root->right = root->left;
                root->left = nullptr;
            } else {
                if (stk.empty()) 
                    return;
                root->right = stk.top();
                stk.pop();
            }
            root = root->right;
        }
    }
};